envoy/formatter/substitution_formatter.h

```diff
@@ -1,31 +1,124 @@
 #pragma once
 
-#include <memory>
-#include <string>
-
-#include "envoy/access_log/access_log.h"
-#include "envoy/common/pure.h"
-#include "envoy/config/typed_config.h"
+#include "envoy/formatter/substitution_formatter_base.h"
 #include "envoy/http/header_map.h"
-#include "envoy/server/factory_context.h"
-#include "envoy/stream_info/stream_info.h"
 
 namespace Envoy {
 namespace Formatter {
 
+/**
+ * HTTP specific substitution formatter context for HTTP access logs or formatters.
+ */
+class HttpFormatterContext {
+public:
+  /**
+   * Constructor that uses the provided request/response headers, response trailers, local reply
+   * body, and access log type. Any of the parameters can be nullptr/empty.
+   *
+   * @param request_headers supplies the request headers.
+   * @param response_headers supplies the response headers.
+   * @param response_trailers supplies the response trailers.
+   * @param local_reply_body supplies the local reply body.
+   * @param log_type supplies the access log type.
+   */
+  HttpFormatterContext(const Http::RequestHeaderMap* request_headers = nullptr,
```

Comment:
Do we need default arguments for this constructor? In general, I prefer to avoid default arguments, unless there is a compelling use case. There might be here, but I'm not sure.

Reply:
The HTTP sustitutuion formatter is reused by lots of modules where have no request/response headers/trailers (like tcp proxy or thrift proxy). 

Their calls looks like this:
```
  return example_formatter->format(*Http::StaticEmptyHeaders::get().request_headers,
                                   *Http::StaticEmptyHeaders::get().response_headers,
                                   *Http::StaticEmptyHeaders::get().response_trailers, stream_info,
                                   absl::string_view(), AccessLog::AccessLogType::NotSet);
```

With the default value in the HttpFormatterContext constructor, we can simplify it to this:

```
  return example_formatter->formatWithContext({}, stream_info);
  // Or this if they has request headers and response headers but no other parameters.
  return example_formatter->formatWithContext({request_headers, response_headers}, stream_info);
```